burn-compute/src/tune/tune_cache.rs

```diff
@@ -19,25 +63,159 @@ pub enum TuneCacheResult<K> {
+            }
         }
+
+        #[cfg(not(feature = "autotune-persistent-cache"))]
+        {
+            if let Some(InMemoryCacheEntry { fastest_index, .. }) = result {
+                return TuneCacheResult::Hit(autotune_operation_set.fastest(*fastest_index));
+            }
+        }
+
         TuneCacheResult::Miss(autotune_operation_set)
     }
 
     pub(crate) fn cache_insert(&mut self, key: K, fastest_index: usize) {
-        self.cache.insert(key, fastest_index);
+        self.in_memory_cache.insert(
+            key,
+            InMemoryCacheEntry {
+                #[cfg(feature = "autotune-persistent-cache")]
+                checksum_checked: true,
+                fastest_index,
+            },
+        );
+    }
+
+    #[cfg(feature = "autotune-persistent-cache")]
+    pub(crate) fn persistent_cache_insert(
+        &mut self,
+        key: K,
+        checksum: String,
+        fastest_index: usize,
+    ) {
+        self.persistent_cache.insert(
+            key,
+            PersistentCacheEntry {
+                checksum,
+                fastest_index,
+            },
+        );
+    }
+
+    /// Load the persistent cache data from disk
+    #[cfg(feature = "autotune-persistent-cache")]
+    pub(crate) fn load(&mut self) -> Result<(), io::Error> {
+        let file_path = self.get_persistent_cache_file_path();
+        // note: reading file from memory is faster than using
+        // serde from_reader with a buffered reader
+        // see issue:
+        // https://github.com/serde-rs/json/issues/160
+        match fs::read_to_string(file_path) {
+            Ok(data) => {
+                let data: Vec<(K, PersistentCacheEntry)> = serde_json::from_str(&data)?;
+                for (key, value) in data.into_iter() {
+                    self.persistent_cache.insert(key, value);
+                }
+                Ok(())
+            }
+            Err(e) => {
+                if e.kind() == std::io::ErrorKind::NotFound {
+                    Ok(())
+                } else {
+                    Err(e)
+                }
+            }
+        }?;
+        for (key, entry) in self.persistent_cache.iter() {
+            self.in_memory_cache.insert(
+                key.clone(),
+                InMemoryCacheEntry {
+                    checksum_checked: false,
+                    fastest_index: entry.fastest_index,
+                },
+            );
+        }
+        Ok(())
+    }
+
+    /// Save the persistent cache on disk
+    #[cfg(feature = "autotune-persistent-cache")]
+    pub(crate) fn save(&self) {
+        let file_path = self.get_persistent_cache_file_path();
+        if let Some(parent_dir) = file_path.parent() {
+            if !parent_dir.exists() {
+                let msg = format!(
+                    "Should be able to create directory '{}' for autotune persistent cache file",
+                    parent_dir.to_str().unwrap()
+                );
```

Comment:
Instead of always formatting the error msg, you can use the following pattern:

```rust
my_function().unwrap_or_else(|| panic!("custom message with format option: {somevariable}"));
```